src/misc/messages.cpp

```diff
@@ -62,49 +83,89 @@ class Message {
 		if (rendered_msg.empty())
 			rendered_msg = convert_ansi_markup(markup_msg.c_str());
 		assert(rendered_msg.length());
-		return rendered_msg.c_str();
+
+		if (!is_utf8_msg)
+			return rendered_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (rendered_cp_msg[cp].empty()) {
+			UTF8_RenderForDos(rendered_msg, rendered_cp_msg[cp], cp);
+			assert(rendered_cp_msg[cp].length());
+		}
+		return rendered_cp_msg[cp].c_str();
 	}
 
-	void Set(const char *markup)
+	void Set(const char *markup, const bool is_utf8)
 	{
 		assert(markup);
-		markup_msg = markup;
+		markup_msg  = markup;
+		is_utf8_msg = is_utf8;
+
 		rendered_msg.clear();
+		markup_cp_msg.clear();
+		rendered_cp_msg.clear();
 	}
 };
 
 static std::unordered_map<std::string, Message> messages;
 static std::deque<std::string> messages_order;
 
-// Add the message if it doesn't exist yet
-void MSG_Add(const char *name, const char *markup_msg)
+static void msg_add(const char *name, const char *markup_msg, const bool is_utf8)
 {
-	const auto &pair = messages.try_emplace(name, markup_msg);
+	const auto &pair = messages.try_emplace(name, markup_msg, is_utf8);
 	if (pair.second) // if the insertion was successful
 		messages_order.emplace_back(name);
 }
 
+// Add the message if it doesn't exist yet
+void MSG_Add(const char *name, const char *markup_msg)
+{
+	msg_add(name, markup_msg, false);
+}
+
 // Replace existing or add if it doesn't exist
-void MSG_Replace(const char *name, const char *markup_msg)
+static void msg_replace(const char *name, const char *markup_msg, const bool is_utf8)
 {
 	auto it = messages.find(name);
 	if (it == messages.end())
-		MSG_Add(name, markup_msg);
+		msg_add(name, markup_msg, is_utf8);
 	else
-		it->second.Set(markup_msg);
+		it->second.Set(markup_msg, is_utf8);
```

Comment:
For these is-utf8 boolean bolt-ons, is that plan that we can easily drop them again once transitions to a pure-utf8 basis?

Or do you envision needing to keep this distinction?

Reply:
If strings in all MSG_Add are all converted to UTF-8, then yes, we may drop these.